diff --git a/utilities.jl b/utilities.jl
index a7e88f7..c363843 100755
--- a/utilities.jl
+++ b/utilities.jl
@@ -106,11 +106,10 @@ function Hk(origlist, sparsity, support)
 	list = real(copy(origlist))
 	ksparse = zeros(length(list))
 	indicesToKeep = (support.==1)
-	dummyvalue = minimum(list)-1
-	list[(support.>-1)] .= dummyvalue
+	candidateInds = findall(support .== -1)
 
-	newIndices = selectperm2(list, sparsity-sum(indicesToKeep))
-	indicesToKeep[newIndices].=true
+	newIndices = selectperm2(list[candidateInds], sparsity-sum(indicesToKeep))
+	indicesToKeep[candidateInds[newIndices]].=true
 
 	ksparse[indicesToKeep]=origlist[indicesToKeep]
 
